### PR TITLE
feat: add `shield:model` command

### PR DESCRIPTION
**Pull request by paulbalandan**
Supersedes and closes #491

### CODE REVIEW

File: docs/concepts.md [hunk 1]
```diff
@@ -25,11 +25,19 @@ on the standard Config class if nothing is found in the database.
 ## User Providers
 
 You can use your own models to handle user persistence. Shield calls this the "User Provider" class. A default model
-is provided for you at `CodeIgniter\Shield\Models\UserModel`. You can change this in the `Config\Auth::$userProvider` setting.
+is provided for you at `CodeIgniter\Shield\Models\UserModel`. You can change this in the `Config\Auth->userProvider` setting.
```
Comment (kenjis) [inline docs/concepts.md:28]: I feel `Config\Auth::$userProvider` is better than `Config\Auth->userProvider`.

Comment (paulbalandan) [reply docs/concepts.md:28]: Wait. I thought I changed this already during the conflict resolution?

Comment (paulbalandan) [reply docs/concepts.md]: Fixed.

**Review by MGatner — COMMENTED** [review]
Renaming it **My**UserModel means that factories won't pick it up automatically when searching for UserModel. This seems a loss to me, maybe not for the library itself but for any supporting code.

**Comment by kenjis** [comment]
`UserModel` instead of `MyUserModel` seems better for the default sample.
There is no need to prefix `My`.

**Comment by kenjis** [comment]
Oh, I get an error if I specify `UserModel`. It's a bug.

File: src/Commands/Generators/Views/usermodel.tpl.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<@php
+
+declare(strict_types=1);
+
+namespace {namespace};
+
+use CodeIgniter\Shield\Models\UserModel;
```
Comment (lonnieezell) [inline src/Commands/Generators/Views/usermodel.tpl.php:7]:
```suggestion
use CodeIgniter\Shield\Models\UserModel as ShieldUserModel;
```

File: src/Commands/Generators/Views/usermodel.tpl.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+class {class} extends UserModel
```
Comment (lonnieezell) [inline src/Commands/Generators/Views/usermodel.tpl.php:9]:
```suggestion
class {class} extends ShieldUserModel
```

File: tests/Commands/UserModelGeneratorTest.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        stream_filter_remove($this->streamFilter);
+        $this->deleteTestFiles();
+    }
+
+    private function deleteTestFiles(): void
+    {
+        $possibleFiles = [
+            APPPATH . 'Models/MyUserModel.php',
+            HOMEPATH . 'src/Models/MyUserModel.php',
```
Comment (lonnieezell) [inline tests/Commands/UserModelGeneratorTest.php:41]:
```suggestion
            APPPATH . 'Models/UserModel.php',
            HOMEPATH . 'src/Models/UserModel.php',
```

**Review by lonnieezell — COMMENTED** [review]
When I tried with UserModel as the generated filename, I think the only issue was the class name conflict. I've made suggestions here to fix that in the code (though not the docs). 

I think this will fix @MGatner concerns, which I agree with. 

Other than that, my only suggestion might be to provide a default classname at the cli prompt.

**Review by MGatner — APPROVED** [review]
I haven't run this but the code looks good. Thanks for making that change.

**Review by lonnieezell — APPROVED** [review]
Looks great. Thanks for taking this one over.

File: src/Commands/Generators/UserModelGenerator.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        $this->component = 'Model';
+        $this->directory = 'Models';
+        $this->template  = 'usermodel.tpl.php';
+
+        $this->classNameLang = 'CLI.generator.className.model';
+        $this->setHasClassName(false);
+
+        $class = $params[0] ?? CLI::getSegment(2) ?? 'UserModel';
+
+        if (in_array(strtolower($class), ['shielduser', 'shieldusermodel'], true)) {
```
Comment (datamweb) [inline src/Commands/Generators/UserModelGenerator.php:68]:
```suggestion
        if (in_array(strtolower($class), ['shielduser', 'shieldusermodel', 'shielduser_model', 'shield_user_model', 'shield_user', 'shield_user_'], true)) {
```
Although nothing should be chosen as the **class name**. Just for note. And probably a few more that I don't know.

**Review by datamweb — APPROVED** [review]

**Comment by datamweb** [comment]
> It's a bug.

@kenjis waiting for you. Did your worries go away?

File: src/Commands/Generators/UserModelGenerator.php (hunk 5 above)
Comment (kenjis) [reply src/Commands/Generators/UserModelGenerator.php:68]: Why do we need `shielduser`? Cannot generate class `ShieldUser`?

Comment (kenjis) [inline src/Commands/Generators/UserModelGenerator.php:68]: Why do we need `shielduser`? Cannot generate class `ShieldUser`?

Comment (kenjis) [reply src/Commands/Generators/UserModelGenerator.php:68]: How about this?

```php
        $classToCheck = CLI::getOption('suffix') ? $class . 'Model' : $class;
        if (strtolower($classToCheck) === 'shieldusermodel') {
            CLI::error('Cannot use `ShieldUserModel` as class name as this conflicts with the parent class.', 'light_gray', 'red');

            return; // @TODO when CI4 is at v4.3+, change this to `return 1;` to signify failing exit
        }
```

Comment (paulbalandan) [reply src/Commands/Generators/UserModelGenerator.php]: Updated. See https://github.com/codeigniter4/shield/pull/558/commits/7e5c4a2069cf570f7e08b934a0f7633c17f7c12e

Comment (kenjis) [reply src/Commands/Generators/UserModelGenerator.php]: Okay, thanks!

**Review by kenjis — APPROVED** [review]
LGTM!

File: src/Commands/Generators/UserModelGenerator.php (hunk 5 above)
Comment (kenjis) [reply src/Commands/Generators/UserModelGenerator.php]: @datamweb If you are okay, please merge this.

Comment (datamweb) [reply src/Commands/Generators/UserModelGenerator.php]: @paulbalandan thank you.
 Really, in many cases, your deep understanding of the code amazes me.

**Review by datamweb — APPROVED** [review]
Thanks all.

**Pull request merged**